File: python/stack/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        answer = []

        def dfs(current, left, right):
            if left + right == n * 2:
                answer.append(current)

            if left > right:
                dfs(current + ")", left, right + 1)

            if left < n:
                dfs(current + "(", left + 1, right)

        dfs("", 0, 0)
        return answer

File: c++/stack/generate-parentheses.cc
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        std::vector<std::string> result;
        dfs(0, 0, "", n, result);
        return result;
    }

private:
    void dfs(int open, int close, std::string s, int n, std::vector<std::string>& result) {
        if (open == close && open + close == n * 2) {
            result.push_back(s);
            return;
        }

        if (open + close == n * 2) {
            return;
        }

        if (open > close) {
            dfs(open, close + 1, s + ")", n, result);
        }

        dfs(open + 1, close, s + "(", n, result);
    }
};
